File: src/android/FirebaseDynamicLinksPlugin.java
package by.chemerisuk.cordova.firebase;

import static by.chemerisuk.cordova.support.ExecutionThread.WORKER;

import android.content.Intent;
import android.net.Uri;
import android.util.Log;

import com.google.firebase.dynamiclinks.DynamicLink;
import com.google.firebase.dynamiclinks.DynamicLink.AndroidParameters;
import com.google.firebase.dynamiclinks.DynamicLink.GoogleAnalyticsParameters;
import com.google.firebase.dynamiclinks.DynamicLink.IosParameters;
import com.google.firebase.dynamiclinks.DynamicLink.ItunesConnectAnalyticsParameters;
import com.google.firebase.dynamiclinks.DynamicLink.NavigationInfoParameters;
import com.google.firebase.dynamiclinks.DynamicLink.SocialMetaTagParameters;
import com.google.firebase.dynamiclinks.FirebaseDynamicLinks;
import com.google.firebase.dynamiclinks.PendingDynamicLinkData;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaArgs;
import org.apache.cordova.PluginResult;
import org.json.JSONException;
import org.json.JSONObject;

import by.chemerisuk.cordova.support.CordovaMethod;
import by.chemerisuk.cordova.support.ReflectiveCordovaPlugin;


public class FirebaseDynamicLinksPlugin extends ReflectiveCordovaPlugin {
    private static final String TAG = "FirebaseDynamicLinks";

    private FirebaseDynamicLinks firebaseDynamicLinks;
    private String domainUriPrefix;
    private CallbackContext dynamicLinkCallback;

    @Override
    protected void pluginInitialize() {
        Log.d(TAG, "Starting Firebase Dynamic Links plugin");

        this.firebaseDynamicLinks = FirebaseDynamicLinks.getInstance();
        this.domainUriPrefix = this.preferences.getString("DOMAIN_URI_PREFIX", "");
    }

    @Override
    public void onNewIntent(Intent intent) {
        if (dynamicLinkCallback != null) {
            respondWithDynamicLink(intent, dynamicLinkCallback);
        }
    }

    @CordovaMethod
    private void getDynamicLink(CallbackContext callbackContext) {
        respondWithDynamicLink(cordova.getActivity().getIntent(), callbackContext);
    }

    @CordovaMethod
    private void onDynamicLink(CallbackContext callbackContext) {
        dynamicLinkCallback = callbackContext;
    }

    @CordovaMethod(WORKER)
    private void createDynamicLink(CordovaArgs args, final CallbackContext callbackContext) throws JSONException {
        JSONObject params = args.getJSONObject(0);
        int linkType = args.getInt(1);
        DynamicLink.Builder builder = createDynamicLinkBuilder(params);
        if (linkType == 0) {
            callbackContext.success(builder.buildDynamicLink().getUri().toString());
        } else {
            builder.buildShortDynamicLink(linkType)
                .addOnCompleteListener(this.cordova.getActivity(), task -> {
                    if (task.isSuccessful()) {
                        callbackContext.success(task.getResult().getShortLink().toString());
                    } else {
                        callbackContext.error(task.getException().getMessage());
                    }
                });
        }
    }

    private void respondWithDynamicLink(Intent intent, final CallbackContext callbackContext) {
        this.firebaseDynamicLinks.getDynamicLink(intent).continueWith(task -> {
            PendingDynamicLinkData data = task.getResult();
            // Hathaway code (Java version) change made to avoid a "phantom" empty deeplink
            // from clobbering a valid deeplink on first launch. Refer to ticket PRG-1613
            if ("".equals(data.getLink())) return null;

            JSONObject result = new JSONObject();
            result.put("deepLink", data.getLink());
            result.put("clickTimestamp", data.getClickTimestamp());
            result.put("minimumAppVersion", data.getMinimumAppVersion());

            if (callbackContext != null) {
                PluginResult pluginResult = new PluginResult(PluginResult.Status.OK, result);
                pluginResult.setKeepCallback(callbackContext == dynamicLinkCallback);
                callbackContext.sendPluginResult(pluginResult);
            }

            return result;
        })
        .addOnFailureListener(e -> {
            if (callbackContext != null && callbackContext != dynamicLinkCallback) {
                callbackContext.sendPluginResult(new PluginResult(PluginResult.Status.OK, (String)null));
            }
        });
    }

    private DynamicLink.Builder createDynamicLinkBuilder(JSONObject params) throws JSONException {
        DynamicLink.Builder builder = this.firebaseDynamicLinks.createDynamicLink();
        builder.setDomainUriPrefix(params.optString("domainUriPrefix", this.domainUriPrefix));
        builder.setLink(Uri.parse(params.getString("link")));
        JSONObject androidInfo = params.optJSONObject("androidInfo");
        if (androidInfo != null) {
            builder.setAndroidParameters(getAndroidParameters(androidInfo));
        }
        JSONObject iosInfo = params.optJSONObject("iosInfo");
        if (iosInfo != null) {
            builder.setIosParameters(getIosParameters(iosInfo));
        }
        JSONObject navigationInfo = params.optJSONObject("navigationInfo");
        if (navigationInfo != null) {
            builder.setNavigationInfoParameters(getNavigationInfoParameters(navigationInfo));
        }
        JSONObject googlePlayAnalyticsInfo = params.optJSONObject("googlePlayAnalytics");
        if (googlePlayAnalyticsInfo != null) {
            builder.setGoogleAnalyticsParameters(getGoogleAnalyticsParameters(googlePlayAnalyticsInfo));
        }
        JSONObject itunesConnectAnalyticsInfo = params.optJSONObject("itunesConnectAnalytics");
        if (itunesConnectAnalyticsInfo != null) {
            builder.setItunesConnectAnalyticsParameters(getItunesConnectAnalyticsParameters(itunesConnectAnalyticsInfo));
        }
        JSONObject socialMetaTagInfo = params.optJSONObject("socialMetaTagInfo");
        if (socialMetaTagInfo != null) {
            builder.setSocialMetaTagParameters(getSocialMetaTagParameters(socialMetaTagInfo));
        }
        return builder;
    }

    private AndroidParameters getAndroidParameters(JSONObject androidInfo) throws JSONException {
        AndroidParameters.Builder androidInfoBuilder;
        if (androidInfo.has("androidPackageName")) {
            androidInfoBuilder = new AndroidParameters.Builder(androidInfo.getString("androidPackageName"));
        } else {
            androidInfoBuilder = new AndroidParameters.Builder();
        }
        if (androidInfo.has("androidFallbackLink")) {
            androidInfoBuilder.setFallbackUrl(Uri.parse(androidInfo.getString("androidFallbackLink")));
        }
        if (androidInfo.has("androidMinPackageVersionCode")) {
            androidInfoBuilder.setMinimumVersion(androidInfo.getInt("androidMinPackageVersionCode"));
        }
        return androidInfoBuilder.build();
    }

    private IosParameters getIosParameters(JSONObject iosInfo) throws JSONException {
        IosParameters.Builder iosInfoBuilder = new IosParameters.Builder(iosInfo.getString("iosBundleId"));
        iosInfoBuilder.setAppStoreId(iosInfo.optString("iosAppStoreId"));
        iosInfoBuilder.setIpadBundleId(iosInfo.optString("iosIpadBundleId"));
        iosInfoBuilder.setMinimumVersion(iosInfo.optString("iosMinPackageVersion"));
        if (iosInfo.has("iosFallbackLink")) {
            iosInfoBuilder.setFallbackUrl(Uri.parse(iosInfo.getString("iosFallbackLink")));
        }
        if (iosInfo.has("iosIpadFallbackLink")) {
            iosInfoBuilder.setIpadFallbackUrl(Uri.parse(iosInfo.getString("iosIpadFallbackLink")));
        }
        return iosInfoBuilder.build();
    }

    private NavigationInfoParameters getNavigationInfoParameters(JSONObject navigationInfo) throws JSONException {
        NavigationInfoParameters.Builder navigationInfoBuilder = new NavigationInfoParameters.Builder();
        if (navigationInfo.has("enableForcedRedirect")) {
            navigationInfoBuilder.setForcedRedirectEnabled(navigationInfo.getBoolean("enableForcedRedirect"));
        }
        return navigationInfoBuilder.build();
    }

    private GoogleAnalyticsParameters getGoogleAnalyticsParameters(JSONObject googlePlayAnalyticsInfo) {
        GoogleAnalyticsParameters.Builder gaInfoBuilder = new GoogleAnalyticsParameters.Builder();
        gaInfoBuilder.setSource(googlePlayAnalyticsInfo.optString("utmSource"));
        gaInfoBuilder.setMedium(googlePlayAnalyticsInfo.optString("utmMedium"));
        gaInfoBuilder.setCampaign(googlePlayAnalyticsInfo.optString("utmCampaign"));
        gaInfoBuilder.setContent(googlePlayAnalyticsInfo.optString("utmContent"));
        gaInfoBuilder.setTerm(googlePlayAnalyticsInfo.optString("utmTerm"));
        return gaInfoBuilder.build();
    }

    private ItunesConnectAnalyticsParameters getItunesConnectAnalyticsParameters(JSONObject itunesConnectAnalyticsInfo) {
        ItunesConnectAnalyticsParameters.Builder iosAnalyticsInfo = new ItunesConnectAnalyticsParameters.Builder();
        iosAnalyticsInfo.setAffiliateToken(itunesConnectAnalyticsInfo.optString("at"));
        iosAnalyticsInfo.setCampaignToken(itunesConnectAnalyticsInfo.optString("ct"));
        iosAnalyticsInfo.setProviderToken(itunesConnectAnalyticsInfo.optString("pt"));
        return iosAnalyticsInfo.build();
    }

    private SocialMetaTagParameters getSocialMetaTagParameters(JSONObject socialMetaTagInfo) throws JSONException {
        SocialMetaTagParameters.Builder socialInfoBuilder = new SocialMetaTagParameters.Builder();
        socialInfoBuilder.setTitle(socialMetaTagInfo.optString("socialTitle"));
        socialInfoBuilder.setDescription(socialMetaTagInfo.optString("socialDescription"));
        if (socialMetaTagInfo.has("socialImageLink")) {
            socialInfoBuilder.setImageUrl(Uri.parse(socialMetaTagInfo.getString("socialImageLink")));
        }
        return socialInfoBuilder.build();
    }
}

File: src/ios/FirebaseDynamicLinksPlugin.m
#import "FirebaseDynamicLinksPlugin.h"

@implementation FirebaseDynamicLinksPlugin

- (void)pluginInitialize {
    NSLog(@"Starting Firebase DynamicLinks plugin");

    if (![FIRApp defaultApp]) {
        [FIRApp configure];
    }

    self.domainUriPrefix = [self.commandDelegate.settings objectForKey:[@"DOMAIN_URI_PREFIX" lowercaseString]];
}

- (void)getDynamicLink:(CDVInvokedUrlCommand *)command {
    CDVPluginResult *pluginResult;
    if (self.lastDynamicLinkData) {
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:self.lastDynamicLinkData];

        self.lastDynamicLinkData = nil;
    } else {
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:nil];
    }
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)onDynamicLink:(CDVInvokedUrlCommand *)command {
    self.dynamicLinkCallbackId = command.callbackId;
}

- (void)createDynamicLink:(CDVInvokedUrlCommand *)command {
    NSDictionary* params = [command.arguments objectAtIndex:0];
    int linkType = [[command.arguments objectAtIndex:1] intValue];
    FIRDynamicLinkComponents *linkBuilder = [self createDynamicLinkBuilder:params];

    if (linkType == 0) {
        CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:linkBuilder.url.absoluteString];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    } else {
        linkBuilder.options = [[FIRDynamicLinkComponentsOptions alloc] init];
        linkBuilder.options.pathLength = linkType;

        [linkBuilder shortenWithCompletion:^(NSURL * _Nullable shortURL,
                                             NSArray<NSString *> * _Nullable warnings,
                                             NSError * _Nullable error) {
            CDVPluginResult *pluginResult;
            if (error) {
                pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:error.localizedDescription];
            } else if (shortURL) {
                pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:shortURL.absoluteString];
            } else {
                pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:nil];
            }
            [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        }];
    }
}

- (FIRDynamicLinkComponents*) createDynamicLinkBuilder:(NSDictionary*) params {
    NSURL* link = [[NSURL alloc] initWithString:params[@"link"]];
    NSString* domainUriPrefix = params[@"domainUriPrefix"];
    if (!domainUriPrefix) {
        domainUriPrefix = self.domainUriPrefix;
    }

    FIRDynamicLinkComponents *linkBuilder = [[FIRDynamicLinkComponents alloc]
                                             initWithLink:link domainURIPrefix:domainUriPrefix];
    NSDictionary* androidInfo = params[@"androidInfo"];
    if (androidInfo) {
        linkBuilder.androidParameters = [self getAndroidParameters:androidInfo];
    }
    NSDictionary* iosInfo = params[@"iosInfo"];
    if (iosInfo) {
        linkBuilder.iOSParameters = [self getIosParameters:iosInfo];
    }
    NSDictionary* navigationInfo = params[@"navigationInfo"];
    if (navigationInfo) {
        linkBuilder.navigationInfoParameters = [self getNavigationInfoParameters:navigationInfo];
    }
    NSDictionary* analyticsInfo = params[@"analyticsInfo"];
    if (analyticsInfo) {
        NSDictionary* googlePlayAnalyticsInfo = params[@"googlePlayAnalytics"];
        if (googlePlayAnalyticsInfo) {
            linkBuilder.analyticsParameters = [self getGoogleAnalyticsParameters:googlePlayAnalyticsInfo];
        }
        NSDictionary* itunesConnectAnalyticsInfo = params[@"itunesConnectAnalytics"];
        if (itunesConnectAnalyticsInfo) {
            linkBuilder.iTunesConnectParameters = [self getItunesConnectAnalyticsParameters:itunesConnectAnalyticsInfo];
        }
    }
    NSDictionary* socialMetaTagInfo = params[@"socialMetaTagInfo"];
    if (socialMetaTagInfo) {
        linkBuilder.socialMetaTagParameters = [self getSocialMetaTagParameters:socialMetaTagInfo];
    }
    return linkBuilder;
}

- (FIRDynamicLinkAndroidParameters*) getAndroidParameters:(NSDictionary*) androidInfo {
    FIRDynamicLinkAndroidParameters* result = [[FIRDynamicLinkAndroidParameters alloc]
                                     initWithPackageName:androidInfo[@"androidPackageName"]];
    NSNumber* minimumVersion = androidInfo[@"androidMinPackageVersionCode"];
    if (minimumVersion) {
        result.minimumVersion = [minimumVersion intValue];
    }
    NSString* androidFallbackLink = androidInfo[@"androidFallbackLink"];
    if (androidFallbackLink) {
        result.fallbackURL = [[NSURL alloc] initWithString:androidFallbackLink];
    }
    return result;
}

- (FIRDynamicLinkIOSParameters*) getIosParameters:(NSDictionary*) iosInfo {
    FIRDynamicLinkIOSParameters* result = [[FIRDynamicLinkIOSParameters alloc]
                                 initWithBundleID:iosInfo[@"iosBundleId"]];
    result.appStoreID = iosInfo[@"iosAppStoreId"];
    result.iPadBundleID = iosInfo[@"iosIpadBundleId"];
    result.minimumAppVersion = iosInfo[@"iosMinPackageVersion"];
    NSString* iosFallbackLink = iosInfo[@"iosFallbackLink"];
    if (iosFallbackLink) {
        result.fallbackURL = [[NSURL alloc] initWithString:iosFallbackLink];
    }
    NSString* iosIpadFallbackLink = iosInfo[@"iosIpadFallbackLink"];
    if (iosIpadFallbackLink) {
        result.iPadFallbackURL = [[NSURL alloc] initWithString:iosIpadFallbackLink];
    }
    return result;
}

- (FIRDynamicLinkNavigationInfoParameters*) getNavigationInfoParameters:(NSDictionary*) navigationInfo {
    FIRDynamicLinkNavigationInfoParameters* result = [[FIRDynamicLinkNavigationInfoParameters alloc] init];
    NSNumber* forcedRedirectEnabled = navigationInfo[@"enableForcedRedirect"];
    if (forcedRedirectEnabled) {
        result.forcedRedirectEnabled = [forcedRedirectEnabled boolValue];
    }
    return result;
}

- (FIRDynamicLinkGoogleAnalyticsParameters*) getGoogleAnalyticsParameters:(NSDictionary*) googlePlayAnalyticsInfo {
    FIRDynamicLinkGoogleAnalyticsParameters* result = [[FIRDynamicLinkGoogleAnalyticsParameters alloc] init];
    result.source = googlePlayAnalyticsInfo[@"utmSource"];
    result.medium = googlePlayAnalyticsInfo[@"utmMedium"];
    result.campaign = googlePlayAnalyticsInfo[@"utmCampaign"];
    result.content = googlePlayAnalyticsInfo[@"utmContent"];
    result.term = googlePlayAnalyticsInfo[@"utmTerm"];
    return result;
}

- (FIRDynamicLinkItunesConnectAnalyticsParameters*) getItunesConnectAnalyticsParameters:(NSDictionary*) itunesConnectAnalyticsInfo {
    FIRDynamicLinkItunesConnectAnalyticsParameters* result = [[FIRDynamicLinkItunesConnectAnalyticsParameters alloc] init];
    result.affiliateToken = itunesConnectAnalyticsInfo[@"at"];
    result.campaignToken = itunesConnectAnalyticsInfo[@"ct"];
    result.providerToken = itunesConnectAnalyticsInfo[@"pt"];
    return result;
}

- (FIRDynamicLinkSocialMetaTagParameters*) getSocialMetaTagParameters:(NSDictionary*) socialMetaTagInfo {
    FIRDynamicLinkSocialMetaTagParameters* result = [[FIRDynamicLinkSocialMetaTagParameters alloc] init];
    result.title = socialMetaTagInfo[@"socialTitle"];
    result.descriptionText = socialMetaTagInfo[@"socialDescription"];
    NSString* socialImageLink = socialMetaTagInfo[@"socialImageLink"];
    if (socialImageLink) {
        result.imageURL = [[NSURL alloc] initWithString:socialImageLink];
    }
    return result;
}

- (void)postDynamicLink:(FIRDynamicLink*) dynamicLink {
    NSMutableDictionary *data = [NSMutableDictionary dictionary];
    NSString* absoluteUrl = dynamicLink.url.absoluteString;
    NSString* minimumAppVersion = dynamicLink.minimumAppVersion;
    BOOL weakConfidence = (dynamicLink.matchType == FIRDLMatchTypeWeak);

    [data setObject:(absoluteUrl ? absoluteUrl : @"") forKey:@"deepLink"];
    [data setObject:(minimumAppVersion ? minimumAppVersion : @"") forKey:@"minimumAppVersion"];
    [data setObject:(weakConfidence ? @"Weak" : @"Strong") forKey:@"matchType"];

    if (self.dynamicLinkCallbackId) {
        CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:data];
        [pluginResult setKeepCallbackAsBool:YES];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:self.dynamicLinkCallbackId];
    } else {
        self.lastDynamicLinkData = data;
    }
}

@end
